### PR TITLE
New package: alps

### PR DESCRIPTION
**Pull request by Sinan81**
A computational physics package.

### CODE REVIEW

File: var/spack/repos/builtin/packages/alps/package.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    # Refs for building from source and recipes
+    # http://alps.comp-phys.org/mediawiki/index.php/Building_ALPS_from_source
+    # https://github.com/easybuilders/easybuild-easyconfigs/tree/master/easybuild/easyconfigs/a/ALPS
+    # https://github.com/conda-forge/alps-feedstock/tree/master/recipe
+
+    # Package failed to build with boost version >= 1.64
+    depends_on('boost@:1.63.0 +chrono +date_time +filesystem +iostreams +mpi +numpy +program_options +python +regex +serialization +system +test +thread +timer')
+    depends_on('fftw')
+    depends_on('hdf5@1.8.17~mpi+hl')
+    depends_on('openblas')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/alps/package.py]: Does `alps` depend strictly on `openblas`? Or:
```python
depends_on('lapack')
```
is a better fit?

File: var/spack/repos/builtin/packages/alps/package.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Alps(CMakePackage):
+    """Algorithms for Physics Simulations
+
+    Keywords: Condensed Matter Physics, Computational Physics
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/alps/package.py]: Keywords are better specified with the `tags` attribute.

File: var/spack/repos/builtin/packages/alps/package.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    # Refs for building from source and recipes
+    # http://alps.comp-phys.org/mediawiki/index.php/Building_ALPS_from_source
+    # https://github.com/easybuilders/easybuild-easyconfigs/tree/master/easybuild/easyconfigs/a/ALPS
+    # https://github.com/conda-forge/alps-feedstock/tree/master/recipe
+
+    # Package failed to build with boost version >= 1.64
+    depends_on('boost@:1.63.0 +chrono +date_time +filesystem +iostreams +mpi +numpy +program_options +python +regex +serialization +system +test +thread +timer')
+    depends_on('fftw')
+    depends_on('hdf5@1.8.17~mpi+hl')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/alps/package.py]: Is the version of `hdf5` strictly fixed?

**Review by alalazo — CHANGES_REQUESTED** [review]
I left a few comments. In general there are instructions [here](http://alps.comp-phys.org/mediawiki/index.php/Building_ALPS_from_source) that don't match what is encoded in the package. Are the build instruction on the official site outdated?

**Comment by Sinan81** [comment]
> I left a few comments. In general there are instructions [here](http://alps.comp-phys.org/mediawiki/index.php/Building_ALPS_from_source) that don't match what is encoded in the package. Are the build instruction on the official site outdated?

Yep! Build instructions are super outdated!

File: var/spack/repos/builtin/packages/alps/package.py (hunk 1 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/alps/package.py]: In Cmake files, there are explicit references to openblas. So, I am guessing we should keep it as it is:
```
BLAS_WORKS:INTERNAL=
//ADVANCED property for variable: BLAS_goto2_LIBRARY
BLAS_goto2_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: BLAS_mkl_LIBRARY
BLAS_mkl_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: BLAS_mkl_em64t_LIBRARY
BLAS_mkl_em64t_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: BLAS_mkl_ia32_LIBRARY
BLAS_mkl_ia32_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: BLAS_mkl_intel_LIBRARY
BLAS_mkl_intel_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: BLAS_mkl_intel_lp64_LIBRARY
BLAS_mkl_intel_lp64_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: BLAS_mkl_rt_LIBRARY
BLAS_mkl_rt_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: BLAS_openblas_LIBRARY
BLAS_openblas_LIBRARY-ADVANCED:INTERNAL=1
BLAS_openblas_WORKS:INTERNAL=1
//Details about finding BLAS
FIND_PACKAGE_MESSAGE_DETAILS_BLAS:INTERNAL=[/mnt/local/spack/opt/spack/linux-centos7-skylake/gcc-6.5.0/openblas-0.3.9-syn6c37toqmrcjfnxph6v5rdvahjsxsj/lib/libopenblas.so][v()]
```

File: var/spack/repos/builtin/packages/alps/package.py (hunk 3 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/alps/package.py]: Turn out not! It is fixed now. Thanks.

File: var/spack/repos/builtin/packages/alps/package.py (hunk 1 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/alps/package.py]: Now I think about it, your suggestion will work. Will push a commit shortly.

**Review by alalazo — APPROVED** [review]

File: var/spack/repos/builtin/packages/alps/package.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    patch('mpi.patch')
+
+    extends('python')
+
+    root_cmakelists_dir = 'alps'
+
+    def cmake_args(self):
+        args = []
+        args.append('Boost_ROOT_DIR=' + self.spec['boost'].prefix)
+        args.append("-DCMAKE_CXX_FLAGS=-std=c++03 ")
```
Comment (tkameyama) [inline var/spack/repos/builtin/packages/alps/package.py]:
```suggestion
        args.append("-DCMAKE_CXX_FLAGS={0}".format(self.compiler.cxx03_flag))
```
But compiler.cxx03_flags is not defined (yet?).

Comment (Sinan81) [reply var/spack/repos/builtin/packages/alps/package.py]: It doesn't look so.

Comment (tkameyama) [reply var/spack/repos/builtin/packages/alps/package.py]: The option for specify the language standard is dependent on compiler and version.
So spack have  some flags to compiler properties.
For example cxx11_flags, cxx14_flags.
But cxx03_flags is not found.

Comment (alalazo) [reply var/spack/repos/builtin/packages/alps/package.py]: I think for most compilers (if not all?) C++98 and C++03 are synonym. Here's an excerpt from GCC man page:
```
c++98
c++03
    The 1998 ISO C++ standard plus the 2003 technical corrigendum and some additional defect reports. Same as -ansi for C++ code.
```

Comment (tkameyama) [reply var/spack/repos/builtin/packages/alps/package.py]: I see.
On spack, cxx98_flags is defined only gcc and fj.
Other compiler is returned null string.

correct code is follows:
```suggestion
        args.append("-DCMAKE_CXX_FLAGS={0}".format(self.compiler.cxx98_flag))
```

**Comment by Sinan81** [comment]
I wonder if we could have used CXX_STANDARD option in CMAKE instead? https://cmake.org/cmake/help/v3.1/prop_tgt/CXX_STANDARD.html

Oh well, the package builds, so let's merge it.

**Pull request merged**